### ISSUE TITLE
Derecho clients must use same CMake target mode (Debug vs Release) as Derecho

### ISSUE THREAD

**Issue by mpmilano** [Derecho-Project/derecho#138]

**Comment by mpmilano** [comment]
We should fix this: that is, we should ensure that the client's choice of debug or release mode for their *own* code does not need to align with their choice of debug or release mode for *derecho's* code.  We are still working out how exactly to make that happen.

One early idea would be to stop using NDEBUG within Derecho itself, and add a Derecho-specific DEBUG flag in its place.  There would be a very large number of places where we need to make this change.  In particular, the mutils libraries would need to make a corresponding change.

**Comment by sagarjha** [comment]
yes, I can't imagine how many things will go wrong if we mixed Debug and Release.

**Comment by songweijia** [comment]
Obviously, mixing application in Debug mode and library in Release mode happens everywhere in software development (Imagine debugging a program.). Libraries like glibc, boost, allows application development with debugging. I think the issue here is that the behavior of derecho code is different in debug mode and release mode. If we align the behaviors in both mode, it should be fine.

**Comment by KenBirman** [comment]
I would suggest that Matthew add a new debug flag specifically for debugging his serialization messages.  Then we avoid this whole issue, because the message format in Derecho would be identical in debug and release mode, and Matthew could simply enable that flag if he is debugging new serialization logic, but this isn't something end-users would normally be doing.

**Comment by mpmilano** [comment]
A more global version of that is what we're going to do --- replace all the
debug flags with a special new one, and then go through and re-enable the
ones that look useful for a user to have control over.

**Comment by etremel** [comment]
Has this been fixed yet? Commit 4c4b577e3d1c8ca9b692c1879efac6ea1037fc82 decoupled the serialization debug code from the standard NDEBUG macro, and it looks like our current version of CMakeLists.txt doesn't rely on the NDEBUG flag (it uses DERECHO_DEBUG and NOLOG instead).

**Comment by mpmilano** [comment]
As of last summer this has been accomplished. But I don't know if the code
added sense then has adhered to it.
